usr.man/man2/ports.2
.\"
.\" $Id: ports.2,v 1.1 1997/02/27 07:32:14 gdr Exp $
.\"
.\" .TH "PORTS IPC" 2 GNO "System Calls" "16 December 1996"
.TH "PORTS IPC" 2 "16 December 1996" GNO "System Calls"
.SH NAME
.BR pbind ,
.BR pcreate ,
.BR pdelete ,
.BR pgetcount ,
.BR pgetport ,
.BR preceive ,
.BR preset ,
.BR psend
\- GNO ports IPC system
.SH SYNOPSIS
.nf
#include <sys/ports.h>

int pcreate (int \fIcount\fR);
int pbind (int \fIportid\fR, const char *\fIname\fR);
int pgetport (const char *\fIname\fR);
int psend (int \fIportid\fR, long \fImsg\fR);
long preceive (int \fIportid\fR);
int pdelete (int \fIportid\fR, int (*\fIdispose\fR)(long));
int preset (int \fIportid\fR, int (*\fIdispose\fR)(long));
int pgetcount (int \fIportid\fR);
.fi
.SH DESCRIPTION
The Ports IPC (interprocess communication) machanism is a very flexible,
powerful, and efficient method of interprocess communication.  A port
is a queue that can contain a number of 32-bit values.  The size of the port
(how many messages it can contain) is specified as the
.IR count
parameter of the
.BR pcreate
call.
.LP
Creation of a port is done with
.BR pcreate .
You must specify the size of the port in this call, which must be at least
1 (one).  The larger the port, the more data it can hold without blocking
the sending process.
.BR pcreate
returns a port ID value that must be used in subsequent calls to the Ports
IPC routines.
.LP
A name may be associated with a port; this allows totally unrelated processes
to access a port without having to communicate the port ID through some
other method, and without knowing the process ID of the other.  To bind a 
name to a port, call 
.BR pbind .
The
.IR name
argument may be any length, but only the first 32 characters are significant.
If a name has already been bound to the chosen
.IR portid ,
-1 is returned and
.BR errno
is set.
.LP
To get the
.IR portid
of a port by it's name, use the 
.BR pgetport
call, with
.IR name
as the name of the port for which you wish to obtain the port ID.
If no port is associated with
.IR name ,
-1 is returned and errno is set.  Names are only unbound from a port
when that port is deleted.
.LP
.BR psend is used to send a 32-bit datum to a port.  If the port is full
(that is, if there are more unread messages in the port than are specified
in the 
.BR pcreate 
call), then the sending process blocks until a message is read from the
port.  Messages are retrieved from a port using the 
.BR preceie
call.
.BR pgetcount
returns the number of messages in the port that have not been received;
this may be used to avoid blocking on a
.BR psend
call.
.LP
If you wish to clear the contents of a port, (for example to synchronize
communication after an error condition), use the 
.BR preset
call.  The arguments to this call are the port ID and the address of a
.IR dispose
function.  Each message in the port, before being cleared, is passed to the
.IR dispose
function so that appropriate clean-up actio nmay be taken on the data.
For example, if the messages correspond to the address of memory blocks
obtained with 
.BR malloc (3),
you could pass
.BR free (3)
as
.IR dispose
to automatically deallocate that memory.  If you don't wish to take any
special action on the data being cleared, pass the NULL pointer for the 
.IR dispose
argument.
.LP
To destroy a port, make the 
.BR pdelete
call.  It accepts the same arguments as 
.BR preset
and they operate as described above.  The difference between
.BR preset
and 
.BR pdelete
is that the latter totally destroys a port; it may no longer be used.
.BR preset 
clears a ports data but leaves the port open for more data transmissions.
.SH EXAMPLES
For an example of the use of ports, see the source code to the print
spooling utilities, 
.BR lpc (1),
.BR lpr (1),
.BR lpd (8),
and
.BR FilePort .
.SH "SEE ALSO"
.BR procsend (2).


kern/gno/ports.c
/*	$Id: ports.c,v 1.1 1998/02/02 08:18:41 taubert Exp $ */

#include "conf.h"
#include "kernel.h"
#include "proc.h"
#include "gno.h"
#include "/lang/orca/libraries/orcacdefs/stdlib.h"
#include "/lang/orca/libraries/orcacdefs/string.h"
#include <sys/errno.h>
#include <sys/ports.h>

#pragma optimize 79

struct	ptnode	*ptfree;	/* list of free queue nodes */
struct	pt	ports[NPORTS];
int	ptnextp;

extern void PANIC(char *s);

#ifdef KERNEL
segment "KERN2     ";
#else
#include "tests/testports.c"
#endif

/*
 *	pinit - initialize all ports
 */

SYSCALL pinit(int maxmsgs)
{
int i;
struct ptnode *next,*prev;

    if ( (ptfree=malloc(maxmsgs*sizeof(struct ptnode)))==NULL )
        PANIC("pinit - insufficient memory");
    for (i = 0; i < NPORTS; i++) {
	ports[i].ptstate = PTFREE;
        ports[i].ptseq = 0;
    }
    ptnextp = NPORTS - 1;

    /* link up free list of message pointer nodes */
    for (prev = next = ptfree; --maxmsgs > 0; prev = next)
	prev->ptnext = ++next;
    prev->ptnext = NULL;
    return(OK);
}

#pragma databank 1
#pragma toolparms 1
/*
 *	pcreate - create a port that allows "count" outstanding messages
 */
pascal SYSCALL KERNpcreate(int count, int *ERRNO)
{
int ps;
int i,p;
struct pt *ptptr;

    if (count < 0) return SYSERR;
    disableps();
    for (i = 0; i < NPORTS; i++) {
	if ((p = ptnextp--) < 0)
	    ptnextp = NPORTS - 1;
        if ((ptptr = &ports[p])->ptstate == PTFREE) {
	    ptptr->ptstate = PTALLOC;
            ptptr->ptname = PTUNNAMED;
            ptptr->ptssem = Kscreate(ERRNO, count);
            if (ptptr->ptssem == SYSERR) {
      bad:      ptptr->ptstate = PTFREE;
                *ERRNO =  ENOMEM;
                enableps();
                return SYSERR;
            }
            ptptr->ptrsem = Kscreate(ERRNO, 0);
	    if (ptptr->ptrsem == SYSERR) {
	        Ksdelete(ERRNO, ptptr->ptssem);
	        goto bad;
            }
	    /* a->i1 = a->i2 = b BROKEN in C 2.0.3 */
#if 0
	    ptptr->pthead = ptptr->pttail = NULL;
#else
	    ptptr->pthead = NULL;
	    ptptr->pttail = NULL;
#endif
            ptptr->ptseq++;
            ptptr->ptmaxcnt = count;
            enableps();
            return p;
        }
    }
    enableps();
    *ERRNO = ENOMEM;
    return SYSERR;
}

/*
 *	psend - send a message to a port by enqueueing it
 */

pascal SYSCALL KERNpsend(int portid, long int msg, int *ERRNO)
{
int ps;
struct pt *ptptr;
int seq;
struct ptnode *freenode;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
            enableps();
            return SYSERR;
    }
    /* wait for space and verify port is still allocated */
    seq = ptptr->ptseq;
    if (commonSwait(ERRNO,ptptr->ptssem,procBLOCKED,BLOCKED_PRECEIVE) == SYSERR) {
	enableps();
	/* *ERRNO set in commonSwait() */
	return SYSERR;
    }
    if (ptptr->ptstate != PTALLOC || ptptr->ptseq != seq) {
	enableps();
	return SYSERR;
    }
    if (ptfree == NULL)
	PANIC("Ports - out of nodes");
    freenode = ptfree;
    ptfree = freenode->ptnext;
    freenode->ptnext = NULL;
    freenode->ptmsg = msg;
    if (ptptr->pttail == NULL) {
	/* a->i1 = a->i2 = b BROKEN in C 2.0.3 */
#if 0
	ptptr->pthead = ptptr->pttail = freenode;
#else
	ptptr->pthead = freenode;
	ptptr->pttail = freenode;
#endif
    } else {
	(ptptr->pttail)->ptnext = freenode;
        ptptr->pttail = freenode;
    }
    Kssignal(ERRNO, ptptr->ptrsem);
    enableps();
    return OK;
}

/*
 *	preceive - receive a message from a port, blocking if port empty
 */

pascal long SYSCALL KERNpreceive(int portid, int *ERRNO)
{
int ps;
struct pt *ptptr;
int seq;
long int msg;
struct ptnode *nxtnode;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
            enableps();
            return SYSERR;
    }
    /* wait for message and verify that the port is still allocated */
    seq = ptptr->ptseq;
    /* sleep, and return EINTR/SYSERR if we were interrupted */
    if (commonSwait(ERRNO,ptptr->ptrsem,procBLOCKED,BLOCKED_PRECEIVE) == SYSERR) {
        enableps();
	/* *ERRNO set in commonSwait() */
        return SYSERR;
    }
    if (ptptr->ptstate != PTALLOC || ptptr->ptseq != seq) {
            enableps();
            return SYSERR;
    }
    /* dequeue first message that is waiting in the port */

    nxtnode = ptptr->pthead;
    msg = nxtnode->ptmsg;
    if (ptptr->pthead == ptptr->pttail) { /* delete last item */
	/* a->i1 = a->i2 = b BROKEN in C 2.0.3 */
#if 0
	ptptr->pthead = ptptr->pttail = NULL;
#else
	ptptr->pthead = NULL;
	ptptr->pttail = NULL;
#endif
    } else ptptr->pthead = nxtnode->ptnext;
    nxtnode->ptnext = ptfree; /* return to free list */
    ptfree = nxtnode;
    Kssignal(ERRNO, ptptr->ptssem);
    enableps();
    return msg;
}

#pragma toolparms 0

/*
 *	_ptclear - used by pdelete and preset to clear a port
 */

_ptclear(struct pt *ptptr, int newstate, int (*dispose)(long int))
{
struct ptnode *p;

    /* put port in limbo until done freeing processes */
    ptptr->ptstate = PTLIMBO;
    ptptr->ptseq++;
    if ((p=ptptr->pthead) != NULL) {
        for (; p != NULL; p=p->ptnext)
            /* only do this if they specified a disposition */
	    if (dispose != NULL) (*dispose)(p->ptmsg);
        (ptptr->pttail)->ptnext = ptfree;
        ptfree = ptptr->pthead;
    }
    if (newstate == PTALLOC) {
	/* a->i1 = a->i2 = b BROKEN in C 2.0.3 */
#if 0
	ptptr->pthead = ptptr->pttail = NULL;
#else
	ptptr->pthead = NULL;
	ptptr->pttail = NULL;
#endif
        /* sreset(ptptr->ptssem, ptptr->ptmaxcnt);
        sreset(ptptr->ptrsem, 0); */
    } else {
        enableps();
        Ksdelete(&errno, ptptr->ptssem);
        Ksdelete(&errno, ptptr->ptrsem);
        disableps();
    }
    ptptr->ptstate = newstate;
}

#pragma toolparms 1
/*
 *	pdelete - delete a port, freeing waiting processes and messages
 */

pascal SYSCALL KERNpdelete(int portid, int (*dispose)(long int), int *ERRNO)
{
int ps;
struct pt *ptptr;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
        enableps();
        return SYSERR;
    }
    _ptclear(ptptr,PTFREE,dispose);
    /* dispose of the port's name */
    if (ptptr->ptname != NULL) {
        free(ptptr->ptname);
        ptptr->ptname = NULL;
    }
    enableps();
    return OK;
}

/*
 *	preset - reset a port, freeing waiting processes and messages
 */

pascal SYSCALL KERNpreset(int portid, int (*dispose)(long int), int *ERRNO)
{
int ps;
struct pt *ptptr;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
        enableps();
        return SYSERR;
    }
    _ptclear(ptptr, PTALLOC, dispose);
    enableps();
    return OK;
}

/*
 *	pbind - binds a port to a name
 */

pascal SYSCALL KERNpbind(int portid, char *name, int *ERRNO)
{
struct pt *ptptr;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
        enableps();
        return SYSERR;
    }
    if (ptptr->ptname != PTUNNAMED) {
        enableps();
        return SYSERR;
    }
    ptptr->ptname = malloc(33);
    strncpy(ptptr->ptname,name,32);
    enableps();
    return OK;
}

pascal SYSCALL KERNpgetport(char *name, int *ERRNO)
{
struct pt *ptptr;
unsigned i;

    disableps();
    for (i = 0; i < NPORTS; i++)
        if ((ports[i].ptstate == PTALLOC) && (ports[i].ptname != NULL)
            && (!strncmp(ports[i].ptname,name,32))) {
            enableps();                      
            return i;
        }
    enableps();
    return SYSERR;
}

pascal SYSCALL KERNpgetcount(int portid, int *ERRNO)
{
struct pt *ptptr;
int c,d;

    disableps();
    if (isbadport(portid) ||
        (ptptr = &ports[portid])->ptstate != PTALLOC) {
        enableps();
        return SYSERR;
    }
    c = Kscount(ERRNO, ptptr->ptssem);
    d = (c == SYSERR) ? ptptr->ptmaxcnt : ptptr->ptmaxcnt - c;
    enableps();
    return d;
}

